### PR TITLE
Complete re-base - Master 2.x

### PR DESCRIPTION
**Pull request by jkandasa**
## Which problem is this PR solving?

- Completely redesigned the performance tests
- Merged existing features and features from @pavolloffay [repository](https://github.com/pavolloffay/jaeger-perf-tests).
- Merged smoke test with this repository
- Now it is possible to run OpenShift Jenkins job from outside OpenShift cluster
- Additional detail: https://issues.jboss.org/browse/TRACING-281

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -12,17 +12,13 @@
...
-  tags:
-    - from:
-        kind: DockerImage
-        name: ${jaeger.agent.image}
-      name: latest
-      referencePolicy:
-        type: Source
+FROM openjdk:alpine
+
+RUN apk --update --no-cache add tar && rm -rf /var/cache/apk/*
```
Comment (pavolloffay) [inline Dockerfile]: why is this needed?

File: src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        public SpansTimer() {
+            if (config.isPerformanceTestLongRunEnabled()) {
+                // actual is 1000 ms, modify maximum available duration based on spans count.
+                // change it to microsecond
+                int maxDuration = 700;
+                if (config.getSpansCount() < 200) {
+                    maxDuration = 700;
+                } else if (config.getSpansCount() < 500) {
+                    maxDuration = 500;
+                } else {
```
Comment (pavolloffay) [inline src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: Not needed the default case can be set at init time

File: src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                span.finish();
+                try {
+                    TimeUnit.MICROSECONDS.sleep(delay);
+                } catch (InterruptedException ex) {
+                    logger.error("exception, ", ex);
+                }
+            } while (count < config.getSpansCount());
+            if (config.isPerformanceTestLongRunEnabled()) {
+                executedCount.incrementAndGet();
+                if (executedCount.get() % 60 == 0 || executedCount.get() == 1) { // print every 1 minute once
```
Comment (pavolloffay) [inline src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: I would rather remove this logs, a lot of unnecessary code with no value.

File: src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    private boolean close;
+    private AtomicInteger executedCount = new AtomicInteger(0);
+
+    private final int expectedCount;
+
+    public CreateSpansRunnable(JaegerTracer tracer, String name, TestConfig config, boolean close) {
+        this.tracer = tracer;
+        this.name = name;
+        this.config = config;
+        this.close = close;
```
Comment (pavolloffay) [inline src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: I think we should move close out of this class. This should only report spans.

File: src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+/**
+ * @author Pavol Loffay
+ */
+@Slf4j
+public class CreateSpansRunnable implements Runnable {
+    class SpansTimer extends TimerTask {
+
+        private long delay = 0;
+
+        public SpansTimer() {
```
Comment (pavolloffay) [inline src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: Not sure why timmer name it seems it only report spans.

Comment (pavolloffay) [reply src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: It's cleaner to have one simple runnable which just reports spans and then reuse it in a new class running as a timer.

File: src/main/java/io/jaegertracing/tests/ISpanCounter.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.jaegertracing.tests;
+
+import java.io.Closeable;
+
+/**
+ * @author Pavol Loffay
```
Comment (pavolloffay) [inline src/main/java/io/jaegertracing/tests/ISpanCounter.java]: Please remove my name from all files which are modified.

File: src/main/java/io/jaegertracing/tests/ISpanCounter.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ */
+public interface ISpanCounter extends Closeable {
```
Comment (pavolloffay) [inline src/main/java/io/jaegertracing/tests/ISpanCounter.java]: Adding I to interfaces is common in C#, not sure if in java.

**Comment by pavolloffay** [comment]
It's hard to review due to size. I have left a couple of comments.

File: Dockerfile (hunk 1 above)
Comment (jkandasa) [reply Dockerfile]: I was using tar command earlier. I do remove this.

File: src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+
+import io.jaegertracing.internal.JaegerTracer;
+import io.jaegertracing.tests.model.TestConfig;
+import io.opentracing.Span;
+import io.opentracing.tag.Tags;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * @author Pavol Loffay
```
[the remark this answers is not in the document]
Comment (jkandasa) [reply src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: I guess It is ok to keep you as the author. Since you are the original creator of these files.
What do you think?

File: src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java (hunk 2 above)
Comment (jkandasa) [reply src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: Thank you! I will update.

File: src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java (hunk 3 above)
Comment (jkandasa) [reply src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: I will increase the printing possibilities bit higher. I use this log to see all the spans are reported within a seconds. This may be useful when we running tests with `Long` running mode.

File: src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+                span.log(logs);
+                span.finish();
+                try {
+                    TimeUnit.MICROSECONDS.sleep(delay);
+                } catch (InterruptedException ex) {
+                    logger.error("exception, ", ex);
+                }
+            } while (count < config.getSpansCount());
+            if (config.isPerformanceTestLongRunEnabled()) {
+                executedCount.incrementAndGet();
```
[the remark this answers is not in the document]
Comment (jkandasa) [reply src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: You mean `SpansTimer` runnable? If yes, Yes, It is running multiple times.

File: src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java (hunk 4 above)
Comment (jkandasa) [reply src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: You mean `close = true` always?

File: src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java (hunk 5 above)
Comment (jkandasa) [reply src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: I will change the class name. I use this class as runnable if we run tests in `Long` running mode. For the quick test, I call `.run()` method.

File: src/main/java/io/jaegertracing/tests/ISpanCounter.java (hunk 7 above)
Comment (jkandasa) [reply src/main/java/io/jaegertracing/tests/ISpanCounter.java]: I'm also not sure about the standard way :)

**Comment by jkandasa** [comment]
@pavolloffay Thank you for your review comments!
I have updated the changes. can you please have a look?

File: src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java (hunk 4 above)
Comment (pavolloffay) [reply src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: No I mean reporter should only report and not close the tracer/reporter at all.

File: src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (pavolloffay) [reply src/main/java/io/jaegertracing/tests/CreateSpansRunnable.java]: If you copy paste a file then yes. But this one contains a very little from the original code. if I was ok with it I would not ask to remove it.

**Comment by pavolloffay** [comment]
@jkandasa is this ready do you want to merge? You can merge if it is approved.

**Comment by jkandasa** [comment]
@pavolloffay Thank you! I do merge this PR now. I will create new PR to split `CreateSpansRunnable` into two classes.
